pkg/util/docker/dockerfile/builder/imageprogress/progress.go

```diff
@@ -132,10 +152,32 @@ type imageProgressWriter struct {
 	layersChangedFn        func(report, report) bool
 }
 
-func (w *imageProgressWriter) readProgress() error {
+func (w *imageProgressWriter) ReadFrom(reader io.Reader) (int64, error) {
+	decoder := json.NewDecoder(reader)
+	return 0, w.readProgress(decoder)
+}
+
+func (w *imageProgressWriter) Write(data []byte) (int, error) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	if w.internalWriter == nil {
+		var pipeIn *io.PipeReader
+		pipeIn, w.internalWriter = io.Pipe()
+		decoder := json.NewDecoder(pipeIn)
+		go func() {
+			err := w.readProgress(decoder)
+			if err != nil {
+				pipeIn.CloseWithError(err)
+			}
+		}()
+	}
+	return w.internalWriter.Write(data)
+}
```

Comment:
Write is implemented because this is after all a io.Writer

However, when you pass the io.Writer to ioutil.Copy, it checks whether the writer implements io.ReaderFrom and if it does, then it uses that interface.

The problem with only implementing io.Writer is that the fsouza/go-dockerclient relies on the result of doing a ioutil.Copy to the progress writer. If the very last thing that comes through is a line with error status { "error": "blah" }, an error never gets reported because all this writer can do is call "CloseWithError" on the input. However, because it was the last thing that was sent, the input is already closed and the error gets lost.